properties/test_index_manipulation.py

```diff
@@ -0,0 +1,273 @@
+                unique=True,
+            )
+        )
+        note(f"> drop_dims: {dims}")
+        self.dataset = self.dataset.drop_dims(dims)
+
+        for dim in dims:
+            if dim in self.indexed_dims:
+                del self.indexed_dims[self.indexed_dims.index(dim)]
+            elif dim in self.multi_indexed_dims:
+                del self.multi_indexed_dims[self.multi_indexed_dims.index(dim)]
+
+    @precondition(lambda self: bool(self.indexed_dims))
+    @rule(data=st.data())
+    def drop_indexes(self, data):
+        self.check_default_indexes = False
+
+        dims = data.draw(
+            st.lists(st.sampled_from(self.indexed_dims), min_size=1, unique=True)
+        )
+        note(f"> drop_indexes: {dims}")
+        self.dataset = self.dataset.drop_indexes(dims)
+
+        for dim in dims:
+            if dim in self.indexed_dims:
+                del self.indexed_dims[self.indexed_dims.index(dim)]
+            elif dim in self.multi_indexed_dims:
+                del self.multi_indexed_dims[self.multi_indexed_dims.index(dim)]
+
+    @property
+    def swappable_dims(self):
+        ds = self.dataset
+        options = []
+        for dim in self.indexed_dims:
+            choices = [
+                name
+                for name, var in ds._variables.items()
+                if var.dims == (dim,)
+                # TODO: Avoid swapping a dimension to itself
+                and name != dim
+            ]
+            options.extend(
+                (a, b) for a, b in itertools.zip_longest((dim,), choices, fillvalue=dim)
+            )
+        return options
+
+    @rule(data=st.data())
+    # TODO: swap_dims is basically all broken if a multiindex is present
+    # TODO: Avoid swapping from Index to a MultiIndex level
+    # TODO: Avoid swapping from MultiIndex to a level of the same MultiIndex
+    # TODO: Avoid swapping when a MultiIndex is present
+    @precondition(lambda self: not bool(self.multi_indexed_dims))
+    @precondition(lambda self: bool(self.swappable_dims))
+    def swap_dims(self, data):
+        ds = self.dataset
+        options = self.swappable_dims
+        dim, to = data.draw(st.sampled_from(options))
+        note(
+            f"> swapping {dim} to {to}, found swappable dims: {options}, all_dims: {tuple(self.dataset.dims)}"
+        )
+        self.dataset = ds.swap_dims({dim: to})
+
+        del self.indexed_dims[self.indexed_dims.index(dim)]
+        self.indexed_dims += [to]
+
+    @invariant()
+    def assert_invariants(self):
+        # note(f"> ===\n\n {self.dataset!r} \n===\n\n")
+        _assert_internal_invariants(self.dataset, self.check_default_indexes)
+
+
+DatasetStateMachine.TestCase.settings = settings(max_examples=300, deadline=None)
+DatasetTest = DatasetStateMachine.TestCase
+
+
+@pytest.mark.skip(reason="failure detected by hypothesis")
+def test_unstack_object():
+    import xarray as xr
+
+    ds = xr.Dataset()
+    ds["0"] = np.array(["", "\x000"], dtype=object)
```

Comment:
It doesn't seem that surprising that you hit bugs related to empty-string or null-prefixed string?  To most C code, those are identical!